torch/csrc/distributed/c10d/comm.cpp

```diff
@@ -85,8 +85,23 @@ void broadcast_coalesced(
   }
 }
 
-PythonCommHook::PythonCommHook(py::object state, py::object hook)
-    : state_(std::move(state)), hook_(std::move(hook)){};
+std::vector<at::Tensor> CommHookInterface::parseFromHookResult(
+    const c10::IValue& result) {
+  TORCH_INTERNAL_ASSERT(
+      result.isPyObject() || result.isTensorList(),
+      "expected the future value is either a PyObject (if PythonCommHook is used) or TensorList (if CppCommHook is used)");
+
+  if (result.isPyObject()) {
+    py::gil_scoped_acquire ag;
+    py::object obj = torch::jit::toPyObject(result);
+    auto value = torch::jit::toIValue(
+        obj, c10::ListType::create(c10::TensorType::get()));
+
+    return value.toTensorVector();
+  }
+
+  return result.toTensorVector();
+}
```

Comment:
This should be a pure virtual method in CommHookInterface such that both PythonCommHook and CppCommHook have their own implementations.

A subclass of CommHookInterface can override runHook to return an IValue of any type (ex: List/Tuple etc.). It's probably best for the subclass to handle the return value and not in the base class itself.